### ISSUE TITLE
Search by case number does not work

### ISSUE THREAD

**Issue by franck-boullier** [unee-t/frontend#679]
# The problem:

When a user enters a case number in the search bar for cases there is no result displayed

# How to replicate the issue:

- In the Dev envo log in as leonel
- enter the case number 1067 (this is an existing, open case that Leonel is able to access)

![image](https://user-images.githubusercontent.com/31331637/50681057-f5d97300-1044-11e9-995b-f42b05efaf4a.png)

## Expected result:

- Case is displayed

## Actual result:

- Case is not visible

![image](https://user-images.githubusercontent.com/31331637/50681011-d3475a00-1044-11e9-97b6-4cd322a9b219.png)

**Comment by PicoCreator** [comment]
Adding a test for this as well : should search also allow search by unit name (not case name/number)?

It was something one of our tester expected, when giving it a try.

**Comment by lironz** [comment]
currently (bug), the search is made on the title only, here are few option to resolve it:

1. a numerical only value will be searched in the id case property, textual value will in the title property - meaning, numerical search will not include matches from title property
2. numerical value will match against both title and id, textual against title only
3. any value, regardless of type, will be matched against a list of properties - currently [id,title] and later maybe more?

**Comment by franck-boullier** [comment]
I'm not sure I understand your option 3 "any value, regardless of type, will be matched against a list of properties - currently [id,title] and later maybe more?"

I'll suggest we do option 4:
- Numerical value will match against Case id only. 
  - If there is a match then the user is send directly to the case page (since there can only be one match)
  - If there is no match we display a message like "There is no OPEN case with this case number"
- Textual against Case title and Unit name

**Comment by nbiton** [comment]
> Numerical value will match against Case id only

This makes less sense IMO. A case's title can contain significant numbers which the user will try to match his search query agains.t.

> If there is a match then the user is sent directly to the case page (since there can only be one match)

As a number is typed one character at a time, and results are shown as the search query is being typed (update after each character), this is not compatible with the current UI behavior and can't be implemented

> Another idea to consider is to leverage the BZ search mechanism for search (and NOT Meteor)

This is far out of  the scope of this bug, and could be considered in a different path in the future.
FYI - the search mechanism has nothing to do with Meteor as a technology, but is implemented as a normal filtering method in plain JS.

**Comment by nbiton** [comment]
(From @lironz 's comment)
> numerical value will match against both title and id, textual against title only

This is the most reasonable approach IMO. The user will not expect his search query to be matched against any other case attribute as far as we can currently tell

**Comment by franck-boullier** [comment]
> this is not compatible with the current UI behavior

Then we should definitely change the current UI behavior: I see no value to a realtime search as the user types in its search criterias TBH...

**Comment by lironz** [comment]
my #3 meant to search any search input in any field we want to be searched

 #4 is clear, except numerical/textual can hit more than once:
 - input "1", matches ids - "1" & "113", same for textual
I we go in the redirection rout, I suggest the auto redirection should happen in a single match situation only, like here: "106" matched only "1067" so user redirected to this single case 1067,
multiple matches will behave the same as currently implemented.

**Comment by franck-boullier** [comment]
> The user will not expect his search query to be matched against any other case attribute as far as we can currently tell

@nbiton: I disagree with that statement actually: user will actually expect that search will also work on the unit name too. 

And user will be happier to discover that search also return cases where a comment is mentioning the search parameters also (more powerful/relevant searches)

**Comment by franck-boullier** [comment]
> my no. 3 meant to search any search input in any field we want to be searched

@lironz gotcha!

> input "1", matches ids - "1" & "113", same for textual

Another reason why we should NOT have real time search as the user is in the middle of typing the search criterias IMO

**Comment by franck-boullier** [comment]
> [leverage the BZ search mechanism for search] This is far out of the scope of this bug

Why? IMO this is 
- Completely within the scope of this bug 
- A much more elegant, powerful and robust approach than a plain text filtering method.

I also don't see any major complexity in implementing this: the API is well documented, easy to use and very similar to the GET BUG method that we use extensively

**Comment by lironz** [comment]
i suggest, currently fixing the bug with:

> (From @lironz 's comment)
> 
> > numerical value will match against both title and id, textual against title only
> 
> This is the most reasonable approach IMO. The user will not expect his search query to be matched against any other case attribute as far as we can currently tell

and open new issue for case search optimization regarding both UI experience and runtime efficiency.

**Comment by lironz** [comment]
the quick fix can also include the addition of unit name field for search

**Comment by nbiton** [comment]
@franck-boullier and I just had a discussion on this issue. Although there are many more features and functionalities we can have the search perform, we can first make minor adjustments to the current UI so it gives what we immediately need, and make it do the rest at a later stage, perhaps under a different search mode (or not).

The adjustments:
* If the user types a number, the search will perform both an id and title filtering, but if "enter" is pressed (or equivalent on mobile) and the number matches a case id, the user will be redirected to the case page.
* "the addition of unit name field for search" IMO that is a good idea too. @franck-boullier do you agree?

**Comment by franck-boullier** [comment]
> The adjustments:
>  - If the user types a number, the search will perform both an id and title filtering, but if "enter" is pressed (or equivalent on mobile) and the number matches a case id, the user will be redirected to the case page.
>  - "the addition of unit name field for search" IMO that is a good idea too. 
>
> @franck-boullier do you agree?

I do thanks
